cloud-controller-manager/do/doks_lb_validator.go

```diff
@@ -0,0 +1,206 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package do
+
+import (
+	"context"
+	"fmt"
+	"github.com/digitalocean/godo"
+	"github.com/go-logr/logr"
+	"golang.org/x/oauth2"
+	v1 "k8s.io/api/core/v1"
+	"net/http"
+	"os"
+	"strconv"
+
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// LBService represents the v1.service lb object data
+type LBService struct {
+	APIVersion string
+	Kind       string
+	Metadata   Metadata
+	Spec       map[string]interface{}
+	Selector   map[string]interface{}
+	Type       string
+}
+
+// Metadata represents the metadata field in the lb service object
+type Metadata struct {
+	Annotations interface{} `json:"annotation data,omitempty"`
+	Name        string
+	Namespace   string
+}
+
+// DOKSLBServiceValidator validates service type LB
+type DOKSLBServiceValidator struct {
+	decoder *admission.Decoder
+	Log     logr.Logger
+	gClient *godo.Client
+}
+
+func initDOClient() (doClient *godo.Client, err error) {
+	token := os.Getenv(doAccessTokenEnv)
+
+	opts := []godo.ClientOpt{}
+
+	opts = append(opts, godo.SetUserAgent("digitalocean-webhook-server/"+version))
+
+	if token == "" {
+		return nil, fmt.Errorf("environment variable %q is required", doAccessTokenEnv)
+	}
+
+	tokenSource := &tokenSource{
+		AccessToken: token,
+	}
+
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	doClient, err = godo.New(oauthClient, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create godo client: %s", err)
+	}
+
+	return doClient, nil
+}
+
+// DOKSLBServiceValidator ...
+func (v *DOKSLBServiceValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	svc := &v1.Service{}
+	v.Log.V(10).Info("decoding received request")
+	err := v.decoder.Decode(req, svc)
+	if err != nil {
+		v.Log.Error(err, "bad request")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	v.Log.V(10).Info("checking received request")
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return admission.Allowed("")
+	}
+
+	// check if request is for LB deletion
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer && svc.DeletionTimestamp != nil {
+		return admission.Allowed("the loadbalancer is deleting")
+	}
+
+	// initialize DO Client if not already initialized
+	if v.gClient == nil {
+		doClient, err := initDOClient()
+		if err != nil {
+			v.Log.Error(err, "failed to initialize DO client")
+			return admission.Errored(http.StatusConflict, err)
+		}
+		v.gClient = doClient
+	}
+
+	clusterID := os.Getenv(doClusterIDEnv)
+	if clusterID == "" {
+		fmt.Println("missing cluster id")
+	}
+	region := "nyc1"
+	if region == "" {
+		fmt.Println("missing region id")
+	}
+
+	nodePools, _, err := v.gClient.Kubernetes.ListNodePools(ctx, clusterID, &godo.ListOptions{})
```

Comment:
This part is problematic for two reasons:

1. It couples the implementation to DOKS again, which we should not do. The upstream portion of CCM fetches the droplet IDs from the Node objects.
2. It ignores filters that the upstream part of the CCM implementation applies, which we don't. Re-implementing them here would be quite tedious and brittle, but not having them means we would have to iterate through possibly hundreds of worker nodes on large clusters. That puts much more stress on the validator and makes it more susceptible to cluster failures.

From our (CCM) perspective, we ideally wouldn't want to include any droplet IDs at all since we only care about validating the non-node customer configuration (i.e., the annotations). Perhaps we can chat with the LB team to see if excluding those (either unconditionally or when a specific flag on the validation request is set) would be possible.

The second best option is to be minimalistic and fetch a single droplet ID through the available Nodes just to satisfy the node portion of the validation request. I personally don't prefer this option since it is more complex (we would need to page through Nodes and stop as soon as we have one with a valid droplet ID / provider ID) and brittle (it does not work when no Nodes with droplet IDs exist).